handler/schedule_test.go

```diff
@@ -150,3 +154,131 @@ func TestScheduleHandler_mergeResourcesWithDefaults(t *testing.T) {
+			schedule:         "@hourly-random",
+			expectedSchedule: "52 * * * *",
+		},
+		{
+			name:             "WhenScheduleRandomDailyGiven_ThenReturnStableRandomizedSchedule",
+			schedule:         "@daily-random",
+			expectedSchedule: "52 4 * * *",
+		},
+		{
+			name:             "WhenScheduleRandomWeeklyGiven_ThenReturnStableRandomizedSchedule",
+			schedule:         "@weekly-random",
+			expectedSchedule: "52 4 * * 4",
+		},
+		{
+			name:             "WhenScheduleRandomMonthlyGiven_ThenReturnStableRandomizedSchedule",
+			schedule:         "@monthly-random",
+			expectedSchedule: "52 4 26 * *",
+		},
+		{
+			name:             "WhenScheduleRandomYearlyGiven_ThenReturnStableRandomizedSchedule",
+			schedule:         "@yearly-random",
+			expectedSchedule: "52 4 26 5 *",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &ScheduleHandler{
+				Config: job.Config{Log: zap.New(zap.UseDevMode(true))},
+			}
+			result := s.randomizeSchedule(name, tt.schedule)
+			assert.Equal(t, tt.expectedSchedule, result)
+		})
+	}
+}
+
+func TestScheduleHandler_getOrGenerateSchedule(t *testing.T) {
+	tests := []struct {
+		name                 string
+		schedule             *v1alpha1.Schedule
+		originalSchedule     string
+		expectedStatusUpdate bool
+		expectedSchedule     string
+	}{
+		{
+			name: "GivenScheduleWithoutStatus_WhenUsingRandomSchedule_ThenPutGeneratedScheduleInStatus",
+			schedule: &v1alpha1.Schedule{
+				Spec: v1alpha1.ScheduleSpec{
+					Backup: &v1alpha1.BackupSchedule{},
+				},
+			},
+			originalSchedule:     "@hourly-random",
+			expectedSchedule:     "26 * * * *",
+			expectedStatusUpdate: true,
+		},
+		{
+			name: "GivenScheduleWithStatus_WhenUsingRandomSchedule_ThenUseGeneratedScheduleFromStatus",
+			schedule: &v1alpha1.Schedule{
+				Spec: v1alpha1.ScheduleSpec{
+					Backup: &v1alpha1.BackupSchedule{},
+				},
+				Status: v1alpha1.ScheduleStatus{
+					GeneratedSchedules: map[v1alpha1.JobType]string{
+						v1alpha1.BackupType: "26 * 3 * *",
+					},
+				},
+			},
+			originalSchedule: "@hourly-random",
+			expectedSchedule: "26 * 3 * *",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &ScheduleHandler{
+				schedule: tt.schedule,
+				Config:   job.Config{Log: zap.New(zap.UseDevMode(true))},
+			}
+			result := s.getOrGenerateSchedule(v1alpha1.BackupType, tt.originalSchedule)
+			assert.Equal(t, tt.expectedSchedule, result)
+			assert.Equal(t, tt.expectedStatusUpdate, s.requireStatusUpdate)
+			if tt.expectedStatusUpdate {
+				assert.Equal(t, tt.expectedSchedule, tt.schedule.Status.GeneratedSchedules[v1alpha1.BackupType])
+			}
+		})
+	}
+}
+
+func TestScheduleHandler_randomizeSchedule_VerifyCronSyntax(t *testing.T) {
+	s := ScheduleHandler{
+		Config: job.Config{Log: zap.New(zap.UseDevMode(true))},
+	}
+	formatString := "%d is not between %d and %d"
+	arr := []int{0, 0, 0, 0, 0}
+	for i := 0; i < 200; i++ {
```

Comment:
Place the `@yearly` and `@weekly` loop into two sub tests functions. It makes test failures easier to read and helps to narrow down what failed.